src/server/scripts/DragonIsles/AzureVault/boss_telash_greywing.cpp

```diff
@@ -0,0 +1,428 @@
+    {
+        Reset();
+        me->DespawnOrUnsummon(5s, 30s);
+        summons.DespawnAll();
+    }
+
+    void JustDied(Unit* /*killer*/) override
+    {
+        Talk(SAY_DEATH);
+        summons.DespawnAll();
+    }
+
+    void MovementInform(uint32 /*type*/, uint32 id) override
+    {
+        if (id == EVENT_JUMP)
+        {
+            if (_isInCenter == false)
+            {
+                DoCastSelf(SPELL_ABSOLUTE_ZERO_DAMAGE);
+                _isInCenter = true;
+            }
+            else if (_isInCenter == true)
+            {
+                DoCastSelf(SPELL_ICE_POWER_PERIODIC);
+                me->RemoveAurasDueToSpell(SPELL_ABSOLUTE_ZERO_SHIELD);
+                me->SetReactState(REACT_AGGRESSIVE);
+                events.ScheduleEvent(EVENT_FROST_BOMBS, 12s + 200ms);
+                events.ScheduleEvent(EVENT_ICY_DEVASTATOR, IsMythicPlus() ? (19s + 600ms) : (23s + 200ms));
+                _isInCenter = false;
+            }
+        }
+    }
+
+    void UpdateAI(uint32 diff) override
+    {
+        scheduler.Update(diff);
+        if (!UpdateVictim())
+            return;
+
+        if (me->HasUnitState(UNIT_STATE_CASTING))
+            return;
+
+        events.Update(diff);
+
+        while (uint32 eventId = events.ExecuteEvent())
+        {
+            switch (eventId)
+            {
+                case EVENT_FROST_BOMBS:
+                    Talk(SAY_FROST_BOMB);
+                    DoCastSelf(SPELL_FROST_BOMB_CAST);
+                    events.ScheduleEvent(EVENT_FROST_BOMBS, 15s);
+                    break;
+                case EVENT_ICY_DEVASTATOR:
+                    if (Unit* target = SelectTarget(SelectTargetMethod::Random, 0, 75.0f, true))
+                    {
+                        Talk(SAY_ICY_DEVASTATOR);
+                        DoCast(target, SPELL_ICY_DEVASTATOR);
+                        events.ScheduleEvent(EVENT_ICY_DEVASTATOR, 23s);
+                    }
+                    break;
+                case EVENT_ABSOLUTE_ZERO:
+                    if (me->GetPower(POWER_ENERGY) >= 100)
+                    {
+                        Talk(SAY_FLY_TO_MIDDLE);
+                        Talk(SAY_ANNOUNCE_ABSOLUTE_ZERO);
+                        me->SetReactState(REACT_PASSIVE);
+                        me->SetDisableGravity(true, true);
+                        me->RemoveAurasDueToSpell(SPELL_ICE_POWER_PERIODIC);
+                        DoCastSelf(SPELL_ABSOLUTE_ZERO_CAST);
+                        DoCastSelf(SPELL_ABSOLUTE_ZERO_SHIELD);
+                        DoCastSelf(SPELL_ACTIVATE_VAULT_RUNE);
+                        events.CancelEvent(EVENT_FROST_BOMBS);
+                        events.CancelEvent(EVENT_ICY_DEVASTATOR);
+                    }
+                    events.Repeat(500ms);
+                    break;
+                default:
+                    break;
+            }
+            if (me->HasUnitState(UNIT_STATE_CASTING))
```

Comment:
Out of curiosity, why are we adding this here again? The boss shouldn't be able to melee at all during casting unless any of the spells have "allow while casting" attributes, especially if we set a return for such before ExecuteEvent.

Reply:
Castcheck is already done implicitly in `DoMeleeAttackIfReady`.

This check is used to prevent weird overlaps which happen at the very same tick. Imagine boss saying textA due to event A, but due to while loop we also execute event B, which triggers textB, but spellcast from event B is prevented due to implicit casting check in DoCast. So we end up printing 2 texts but have only one cast.

Another way to prevent this would be using `if (... = events.ExecuteEvent())` instead of using a loop

Reply:
So the issue is entirely made out of our choice to use a while to use ExecuteEvents, which also prevents having certain update ticks to ignore next call?